### PR TITLE
Speedup Hibernate ORM's enhancement of large models

### PR DESCRIPTION
**Pull request by Sanne**
This works far better, in my local tests.

I feel it's ready, but it will require a fix on the ORM side to be merged first:
 - https://github.com/hibernate/hibernate-orm/pull/8263

This brings enhancement times of the large model provided as benchmark down to ~1 second, without enabling Quarkus's caching capabilities of caching enhancement of unmodified classes (which we should also enable, but it's a separate aspect)

### CODE REVIEW

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/integration/QuarkusClassFileLocator.java
Comment (yrodiere) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/integration/QuarkusClassFileLocator.java:9]: Maybe a short explanation of why this class is better than the default impl would help? Let's think of the poor souls having to handle bugfixes after this gets merged :D

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/integration/QuarkusEnhancementContext.java
Comment (yrodiere) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/integration/QuarkusEnhancementContext.java:24]: Why are we doing this, if it doesn't matter?

**Review by yrodiere — COMMENTED** [review]
Thanks!

Seems safe to me, and it's great if it speeds things up so much, but could you please add a few comments? I'm afraid it's not as obvious for us as it is for you :)

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/integration/QuarkusEnhancementContext.java
Comment (Sanne) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/integration/QuarkusEnhancementContext.java:24]: good point. I guess it's a remnant of previous attempt, in which I had a custom one.. but it's also good to have a place to comment about it.
I'll replace it with throwing an exception, so to catch future misunderstandings.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/integration/QuarkusClassFileLocator.java
Comment (Sanne) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/integration/QuarkusClassFileLocator.java:9]: ah right. Yes in fairness this isn't better than the standard `ClassFileLocator` yet, as it's exactly the same actually except not needing the complicated custom classloader to defer to.
But I also have plans to make it much better in a subsequent commits so to get rid of the expansion logic that creates multiple `byte[]` - IFF it all works: I'd like to know if it's worth doing that first.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java
Comment (Sanne) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java:37]: P.S. we could add some more "framework types" here. e.g. Panache might be worth it, perhaps Jackson annotations - what else do people frequently put on entities?

**Comment by Sanne** [comment]
Updated to add the suggested comments; also now registering the Panache super types as "core types" so they'd also benefit from caching, if used.

**Comment by Sanne** [comment]
> also now registering the Panache super types as "core types" so they'd also benefit from caching, if used.

Note to self: this implies that if we have live-reload tests in Panache which have entities using the same package name as Panache itself they might fail now for the silly reason of us skipping the work on such classes.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java
Comment (yrodiere) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java:40]: > > also now registering the Panache super types as "core types" so they'd also benefit from caching, if used.
> 
> Note to self: this implies that if we have live-reload tests in Panache which have entities using the same package name as Panache itself they might fail now for the silly reason of us skipping the work on such classes.

It also implies that demos using the `org.hibernate` package will no longer have live-reload... That's about every single demo the Hibernate team worked on. Certainly every single demo of mine.

Can't this list be generated at build time instead? So that we make extra sure not to include any application classes in the "reused" pool?

Comment (yrodiere) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java:40]: > It also implies that demos using the `org.hibernate` package will no longer have live-reload... That's about every single demo the Hibernate team worked on. Certainly every single demo of mine.

It may also affect test case templates, at least if you port this stuff to Hibernate ORM...

Package-based checks really seem like they could bring a lot of trouble when working with the community (demos/bug reports) :/

Comment (Sanne) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java:40]: I agree and that's why the list is not hardcoded..  in Hibernate ORM I've refrained from adding the `org.hibernate` package to avoid such troubles.

 - https://github.com/hibernate/hibernate-orm/pull/8263/files#diff-ec2d0f0c23fe593565384ac5681d0a4873b6bc7d06f44157eb9d02b81c394dabR39

We can do the same here if you prefer..  I just don't see it very likely that people will be using the Panache namespace to create their own applications. Besides, the rule of "we should really not materialize this class" is for WildFly but I'm not sure if it would cause any trouble for Quarkus, I'm avoiding to materialize application classes out of excess of caution but I don't think it matters in Quarkus specifically - which is part of why I've added this.

But I'd prefer to be cautious indeed, if you're still concerned after that ^ I'm happy to change it.

Comment (yrodiere) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java:40]: > I'm not sure if it would cause any trouble for Quarkus, I'm avoiding to materialize application classes out of excess of caution but I don't think it matters in Quarkus specifically

It probably matters for our own test suite, given the various metaspace (classloader) leaks we still have in `@QuarkusUnitTest`. In the real world, I don't know for sure, but we've had classloader leaks on live-reload in the past, so I wouldn't be surprised if we still had some.

And of course there are probably other concerns than just memory leaks, and I know even less about these.

> We can do the same here if you prefer..

I would. I'd argue that even caching the `jakarta` package could mean trouble in some cases, but I'll grant you that's going a bit far :)

In the case of Quarkus we know precisely which classes are application classes, which is I was suggesting to build this knowledge in Quarkus' own core pool.

But now that I think about it, I wonder if Quarkus' own knowledge is really necessary. The persistence unit descriptors do already include a list of model package/classes, don't they? And IIRC we absolutely need all model classes to be in there, or there could be malfunctions in other areas unrelated to bytecode enhancement. In the case of Quarkus, I think that classes that are not listed there aren't even enhanced.
So... couldn't you rely on that and use it as a "denylist" instead of the current "allowlist"?

See:

https://github.com/quarkusio/quarkus/blob/ab26345d9f689b4785e27754eb088b19b2347bb5/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaModelBuildItem.java#L75-L80

https://github.com/quarkusio/quarkus/blob/469fd3ae2e670dc7f43224242bb234edc591caf1/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java#L887-L897

Comment (Sanne) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java:40]: This isn't just about the model classes. Let me explain: what I observed during profiling is that when an entity class is introspected by ByteBuddy as it's building its model, it also describes all symbols referred to it:
 - the `@Entity` annotation on the entity makes it look for the class of the Entity annotation
 - a property having "String" as type makes it look for the String class
 - a propery having a "Map" makes it look for Map..
 - any other interface, annotation, etc..  on the mapped type or any signature of its methods and fields.

That's a lot of loading and bytecode parsing for each entity we're enhancing, as (before this patch) they would be repeated for each entity being enhanced.
After lookups their bytearray is parsed and BB extracts the datastructure it needs from them - this is what I'm caching, to not have to re-parse the structure of common types such as `@Entity` , "@Id" thousands of times - and record each of them in internal maps taking memory and making lookups heavy. So primarily JPA annotations, but one can see there's plenty other interesting annotations e.g. `jakarta.annotations.Generated` is hit fairly common.

The observation is that nobody is going to *modify* the bytecode @Entity annotation - and if they did, in this particular case livereload would actually still work as what ByteBuddy needs here about such objects is just to make sure the exist and are valid to be placed in a certain place.

So I included the Panache super class, because (if using Panache) you're most likely going to need that symbol for each and every entity again.

In summary, the "deny" list wouldn't include these, making it a little more dangerous.

On the other hand, the cache refactoring is already quite effective even w/o adding "dangerous" types to the CoreTypePool - they are effective even adding them to the ModelTypePool : the difference is that Quarkus is holding the CoreTypePool cache across the whole lifecycle of the JVM in dev mode, while we create a fresh instance of `ModelTypePool` for each live-reload - so we're still benefitting from the cache of ALL types during enhancement of N entities, but we throw it away and restart on the next live-reload.

So I'd suggest: let's limit it to more precise package names only?  I'd be fine to not cache `jakarta.annotations.Generated` but do cache the `jakarta.persistence` and `org.hibernate.annotations`.

If people then have problems because their app is using these packages I believe they deserve it :)

Comment (yrodiere) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java:40]: > In summary, the "deny" list wouldn't include these, making it a little more dangerous.

I'm not sure what "these" is referring to exactly, but if it's the Panache super classes, they actually should be included in the list of classes of each persistence unit.

Anyway, that's a moot point, because:

> So I'd suggest: let's limit it to more precise package names only? I'd be fine to not cache `jakarta.annotations.Generated` but do cache the `jakarta.persistence` and `org.hibernate.annotations`.
> 
> If people then have problems because their app is using these packages I believe they deserve it :)

:+1: seems safe enough

Comment (Sanne) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java:40]: I think I'll go with:
```
"jakarta.persistence.", "java.", "org.hibernate.annotations.",                                                                    "io.quarkus.hibernate.reactive.panache.", "io.quarkus.hibernate.orm.panache",  "org.hibernate.search.mapper.pojo.mapping.definition.annotation", "jakarta.validation.constraints"
```

Any more you'd recommend?

It doesn't have to be comprehensive, so balancing things. Also don't want to have to check a long list for each loaded type , if there is no high likelyhood for it to pay back with substantial benefits.

Comment (Sanne) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java:40]: > "org.hibernate.search.mapper.pojo.mapping.definition.annotation"

And I wish this was shorter :P

Comment (yrodiere) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java:40]: Amen to this. We all make mistakes.

**Comment by Sanne** [comment]
We decided to use:
"jakarta.persistence.", "java.", "org.hibernate.annotations.", "io.quarkus.hibernate.reactive.panache.", "io.quarkus.hibernate.orm.panache.", "org.hibernate.search.mapper.pojo.mapping.definition.annotation.", "jakarta.validation.constraints."

**Comment by Sanne** [comment]
Rebased: simplified implementation taking advantage of eb69d59697e395db49e8e9398d6a80fd1657892f

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/integration/QuarkusClassFileLocator.java
Comment (yrodiere) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/integration/QuarkusClassFileLocator.java:17]: I think it's implemented now, isn't it what this `IoUtil` is about?

**Review by yrodiere — APPROVED** [review]
Thanks. LGTM, just a small doubt that there might be an obsolete comment.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/integration/QuarkusClassFileLocator.java
Comment (Sanne) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/integration/QuarkusClassFileLocator.java:17]: Not quite..  `IoUtil` delegates to the stream implementation, and some of them are very inefficient especially when loading a stram from a compressed jar. Luckily an ORM entity is not frequently stored in jars.

That's where my other classloader changes come into play.. but not all pieces have come together yet.

**Comment by Sanne** [comment]
Please don't merge - I've pushed an experiment and need CI feedback.

**Comment by Sanne** [comment]
@yrodiere I understand what's going on now, unfortunatley sharing the ByteBuddy caches would require much more work - and is tricky.

So I suggest we merge this version: it shares the caches of "core types" (which we know are not being modified) but isolates the caches used for bytecode enhancement in threadlocals to ensure independent interactions.

It's taking the safer approach - there's room for improvement as this forces each thread to repeat quite some work, and also allocate large structures w/o sharing those, but that would require much more work to share them safely, so I'd rather take the 80% improvement now and keep it safe until we have time to add more stricter integration tests.

Ok to merge? I'd hit the button as it's already approved, but since several changes were made...

**Comment by mapuci** [comment]
offtopic:

@Sanne could you please take a look at this issue? It is kindof related to what you are doing here :)
https://github.com/quarkusio/quarkus/issues/40719
There is something wrong with enhanced classes or even their classloaders...(in quarkus 2 and 3, but not in wildfly).
Reproducer repo contains enhanced .class files, so you can take a look at them.

I tried to reproduce problem with this branch, there is still a bug.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java
Comment (yrodiere) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java:138]: You could clear the thread-local? Maybe that's not the best place to do this, but it would be nice to at least try... somewhere...

We've been bitten by classloader leaks caused by threadlocals in the past... And it can get quite bad.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java
Comment (yrodiere) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java:114]: This threadlocal isn't `static` -- is this on purpose?

From what I can see `ThreadsafeLocator` is, in practice, a singleton, so using `static` here would make threadlocal cleanup easier. You know, where we restart in dev mode, drop half the classloaders but keep the same threads, resulting in threadlocals piling up...

**Review by yrodiere — COMMENTED** [review]
I'm really not sure using threadlocals without ever clearing them is safe in dev mode, but I'll let you check that, and merge if you are confident.

**Comment by Sanne** [comment]
Pushed a rebase, as I plan to resume this work this weekend.

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java
Comment (Sanne) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java:114]: I would prefer to avoid the `static` actually as that's what makes the Threadlocal depend on the classloader, and therefore causes a whole cathegory of such issues.

To make sure it's not causing a circular reference via threads to classloaders I now manually insepcted it via MAT after performing some cycles of live reload, and it seems fine.

I also more aggreesively clear the threadlocal now - it wasn't strictly necessary but it won't hurt to be cautious, thanks for making me think twice about that :+1:

File: extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java
Comment (gsmet) [inline extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java:36]: Shouldn't we use `JAVA_V17` now? (line just below)

I would create a separate PR to do it but it's going to be in your way so I will let you address it :).

Comment (Sanne) [reply extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java:36]: good point :)
Amended

**Comment by dashorst** [comment]
Nice work @Sanne , looking forward to trying this out in a next quarkus release!

**Pull request merged**
